bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/overlay/HistoryTextWrapper.java

```diff
@@ -0,0 +1,207 @@
+import org.eclipse.swt.events.DisposeListener;
+import org.eclipse.swt.events.FocusListener;
+import org.eclipse.swt.events.KeyListener;
+import org.eclipse.swt.events.ModifyListener;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.graphics.Color;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Table;
+import org.eclipse.swt.widgets.TableItem;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.swt.widgets.ToolItem;
+import org.eclipse.swt.widgets.Widget;
+
+import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.layout.GridLayoutFactory;
+
+import org.eclipse.ui.internal.findandreplace.FindReplaceMessages;
+import org.eclipse.ui.internal.findandreplace.HistoryStore;
+
+/**
+ * Wrap a Text Bar and a ToolItem to add an input history. The text is only
+ * stored to history when requested by the Client code, the history is stored in
+ * a {@code HistoryStore} provided by the client. The history bar behaves like a
+ * normal {@code Text}.
+ */
+public class HistoryTextWrapper extends Composite {
+	private Text textBar;
+	private AccessibleToolBar tools;
+	private ToolItem dropDown;
+	private SearchHistoryMenu menu;
+	private HistoryStore history;
+
+	public HistoryTextWrapper(HistoryStore history, Composite parent, int style) {
+		super(parent, SWT.NONE);
+		GridLayoutFactory.fillDefaults().numColumns(2).applyTo(this);
+
+		this.history = history;
+
+		textBar = new Text(this, style);
+		GridDataFactory.fillDefaults().grab(true, true).align(GridData.FILL, GridData.CENTER).applyTo(textBar);
+		tools = new AccessibleToolBar(this);
+		dropDown = new AccessibleToolItemBuilder(tools).withStyleBits(SWT.PUSH)
+				.withToolTipText(FindReplaceMessages.FindReplaceOverlay_searchHistory_toolTip)
+				.withImage(FindReplaceOverlayImages.get(FindReplaceOverlayImages.KEY_OPEN_HISTORY))
+				.withSelectionListener(SelectionListener.widgetSelectedAdapter(e -> createHistoryMenuDropdown()))
+				.build();
+
+		listenForKeyboardHistoryNavigation();
+	}
+
+	private void listenForKeyboardHistoryNavigation() {
+		addKeyListener(KeyListener.keyPressedAdapter(e -> {
+			if (e.keyCode == SWT.ARROW_UP || e.keyCode == SWT.ARROW_DOWN) {
+				int stepDirection = e.keyCode == SWT.ARROW_UP ? 1 : -1;
+				navigateInHistory(stepDirection);
+			}
+		}));
+	}
+
+	private void createHistoryMenuDropdown() {
+		if (menu != null && okayToUse(menu.getShell()) || !dropDown.isEnabled()) {
+			return;
+		}
+
+		dropDown.setEnabled(false);
+		menu = new SearchHistoryMenu(getShell(), history, SelectionListener.widgetSelectedAdapter(f -> {
+			Table table = (Table) f.widget;
+			TableItem[] selection = table.getSelection();
+			if (selection.length == 0) {
+				return;
+			}
+			String text = selection[0].getText();
+			if (text != null) {
+				textBar.setText(text);
+			}
+		}));
```

Comment:
OO design: It would be cleaner to have the `SearchHistoryMenu` accept some `SelectedItemChangedListener`, being a functional interface for a consumer of `TableItem` (the one that changed). The current implementation assumes that the the `SearchHistoryMenu` attaches the given listener to a table, which introduces a dependency to implementation details of the `SearchHistoryMenu`.

Reply:
I agree! I have created a `bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/overlay/HistoryItemSelectedListener.java` which abstracts from this to the outside user